src/gui/clips/SampleClipView.cpp

```diff
@@ -205,44 +208,50 @@ void SampleClipView::paintEvent( QPaintEvent * pe )
 {
 	QPainter painter( this );
 
-	if( !needsUpdate() )
+	const auto re = pe->rect();
+	const auto region = QRect(re.x(), rect().y(), re.width(), rect().height());
+
+	if (m_paintPixmap.isNull() || m_paintPixmap.size() != size())
 	{
-		painter.drawPixmap( 0, 0, m_paintPixmap );
-		return;
+		m_paintPixmap = QPixmap(size());
 	}
 
-	setNeedsUpdate( false );
+	bool muted = m_clip->getTrack()->isMuted() || m_clip->isMuted();
+	bool selected = isSelected();
 
-	if (m_paintPixmap.isNull() || m_paintPixmap.size() != size())
+	if (!needsUpdate())
 	{
-		m_paintPixmap = QPixmap(size());
+		painter.drawPixmap(region, m_paintPixmap, region);
+		return;
 	}
 
+	setNeedsUpdate(false);
+
 	QPainter p( &m_paintPixmap );
+	p.setClipRegion(region);
 
-	bool muted = m_clip->getTrack()->isMuted() || m_clip->isMuted();
-	bool selected = isSelected();
+	QColor c = painter.background().color();
 
 	QLinearGradient lingrad(0, 0, 0, height());
-	QColor c = painter.background().color();
 	if (muted) { c = c.darker(150); }
 	if (selected) { c = c.darker(150); }
 
 	lingrad.setColorAt( 1, c.darker( 300 ) );
 	lingrad.setColorAt( 0, c );
 
 	// paint a black rectangle under the clip to prevent glitches with transparent backgrounds
-	p.fillRect( rect(), QColor( 0, 0, 0 ) );
+	p.fillRect( region, QColor( 0, 0, 0 ) );
```

Comment:
Do you still want to split the partial rendering changes made here into another PR?

Reply:
Still not sure if I should undo the changes here. Let's first copy the changes here into another PR and continue there. I'll try not to change the related stuff in this PR. We'll remove the old code once the other PR is complete and merged. What do you think?

Reply:
Any waveform caching related changes in this function I am okay with: they are in scope for this PR. But the changes to use `region` instead of `rect()` look like the partial rendering feature that's specific to `SampleClipView`, which can probably be branched out.

And yeah, it can be here until the new PR is made.